presto-matching/src/main/java/com/facebook/presto/matching/Property.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.matching;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public interface Property<F, T>
```

Comment:
It is a `Property` :) It's purpose (along with `Pattern` and `PropertyPattern`) is to build a DSL so composing methods are to be expected. The interface is implemented once, in the `optionalProperty` method.

> Then PropertyPattern could be renamed to just Pattern and WithPattern to PropertyPattern.

Not all Pattern-s are PropertyPatterns, so the rename would be incorrect (or we'd need an other, better name for Pattern, which is hard to imagine :)). I'm thinking of a way to get rid of `PropertyPattern`, but don't have a good answer so far.